libs/pageserver_api/src/shard.rs

```diff
@@ -0,0 +1,132 @@
+const DEFAULT_STRIPE_SIZE: ShardStripeSize = ShardStripeSize(256 * 1024 / 8);
+
+/// The ShardIdentity contains the information needed for one member of map
+/// to resolve a key to a shard, and then check whether that shard is ==self.
+#[derive(Clone, Copy)]
+struct ShardIdentity {
+    layout: ShardLayout,
+    number: ShardNumber,
+    count: ShardCount,
+    stripe_size: ShardStripeSize,
+}
+
+/// The location of a shard contains both the logical identity of the pageserver
+/// holding it (control plane's perspective), and the physical page service port
+/// that postgres should use (endpoint's perspective).
+#[derive(Clone)]
+struct ShardLocation {
+    id: NodeId,
+    page_service: (url::Host, u16),
+}
+
+/// The ShardMap is sufficient information to map any Key to the page service
+/// which should store it.
+#[derive(Clone)]
+struct ShardMap {
+    layout: ShardLayout,
+    count: ShardCount,
+    stripe_size: ShardStripeSize,
+    pageservers: Vec<Option<ShardLocation>>,
+}
+
+impl ShardMap {
+    pub fn get_location(&self, shard_number: ShardNumber) -> &Option<ShardLocation> {
+        assert!(shard_number.within_count(self.count));
+        self.pageservers.get(shard_number.0 as usize).unwrap()
+    }
+
+    pub fn get_identity(&self, shard_number: ShardNumber) -> ShardIdentity {
+        assert!(shard_number.within_count(self.count));
+        ShardIdentity {
+            layout: self.layout,
+            number: shard_number,
+            count: self.count,
+            stripe_size: self.stripe_size,
+        }
+    }
+
+    pub fn get_shard_number(&self, key: &Key) -> ShardNumber {
+        key_to_shard_number(self.count, self.stripe_size, key)
+    }
+
+    pub fn default_with_shards(shard_count: ShardCount) -> Self {
+        ShardMap {
+            layout: LAYOUT_V1,
+            count: shard_count,
+            stripe_size: DEFAULT_STRIPE_SIZE,
+            pageservers: (0..shard_count.0 as usize).map(|_| None).collect(),
+        }
+    }
+}
+
+impl ShardIdentity {
+    pub fn get_shard_number(&self, key: &Key) -> ShardNumber {
+        key_to_shard_number(self.count, self.stripe_size, key)
+    }
+}
+
+impl Default for ShardIdentity {
+    /// The default identity is to be the only shard for a tenant, i.e. the legacy
+    /// pre-sharding case.
+    fn default() -> Self {
+        ShardIdentity {
+            layout: LAYOUT_V1,
+            number: ShardNumber(0),
+            count: ShardCount(1),
+            stripe_size: DEFAULT_STRIPE_SIZE,
+        }
+    }
+}
+
+/// Where a Key is to be distributed across shards, select the shard.  This function
+/// does not account for keys that should be broadcast across shards.
+fn key_to_shard_number(count: ShardCount, stripe_size: ShardStripeSize, key: &Key) -> ShardNumber {
+    // Fast path for un-sharded tenants
+    if count == ShardCount(0) {
+        return ShardNumber(0);
+    }
+
+    let mut hasher = mur3::Hasher32::with_seed(0);
```

Comment:
Is murmur faster than xxhash3 for these 104 bit inputs or is there some other reason to use it? I vaguely recall the 32 and 64 bit variants of it produced different values.

The crate is 0.1.0 -- I guess this will become a testability burden that we have a hash function which produces the same values across versions, should we consider adopting it within codebase?

(Did not read the latest on the RFC if we are somehow able to evolve the sharding function -- apologies for the noise if so.)

Reply:
murmur was suggested as something convenient to use from the postgres side, although not with a strong preference: https://neondb.slack.com/archives/C03QLRH7PPD/p1696933015237419?thread_ts=1696932827.671149&cid=C03QLRH7PPD

> The crate is 0.1.0 -- I guess this will become a testability burden that we have a hash function which produces the same values across versions, should we consider adopting it within codebase?

Yeah, we should consider this before finalizing.  The code in `shard.rs` is mainly meant to just make the key->stripe->shard mapping very explicit.

> (Did not read the latest on the RFC if we are somehow able to evolve the sharding function -- apologies for the noise if so.)

The layout version field is meant to allow for that, although I don't anticipate needing it in the near term.

Reply:
>Yeah, we should consider this before finalizing. The code in shard.rs is mainly meant to just make the key->stripe->shard mapping very explicit.

For completeness: we ended up using a hand-crafted murmur3 rust function, to ensure it exactly matches the postgres C impl.